### DISCUSSION TITLE
how to accelerate the rendering of VTK and transformation of data from OCCT to VTK

### DISCUSSION THREAD

**Discussion by newguyyy** [community/community#137115] [Programming Help]
### Body

Each TopoDS_Shape is a small ball created by the BRepPrimAPI_MakeSphere, and currently requires VTK to visualize all the balls
However, it takes a long time to convert data from OCCT to VTK and how the visualization time can be optimized
At present, multi-threading is used for data processing, but the optimization effect is limited, and it currently takes about 5000ms to visualize 1000 small balls, and we want to further optimize it.
```
TopTools_SequenceOfShape SeqShape = ballArray(face);
 //多线程性能优化
 Standard_Integer numTreads = thread::hardware_concurrency();
 Standard_Integer shapesPerthread = SeqShape.Length() / numTreads + 1;

 vector<thread> threads;
 for (Standard_Integer i = 0; i < numTreads; i++)
 {
     TopTools_SequenceOfShape subsets;
     for (Standard_Integer j = i * shapesPerthread; j < (i + 1) * shapesPerthread && j < SeqShape.Length(); j++)
     {
         subsets.Append(SeqShape(j + 1));
     }
     threads.emplace_back(ProcessShape, subsets, newRen.Get());
 }
 // Ensure all threads are joined
 for (auto& thread : threads) {
     if (thread.joinable()) {
         thread.join();
     }
 }
```
This is the function that is called by multithreading

```
void ProcessShape(const TopTools_SequenceOfShape& shapes, vtkRenderer* render)
{
	//vector<vtkSmartPointer<vtkActor>> actors;
	TopoDS_Compound compound;
	BRep_Builder builder;
	builder.MakeCompound(compound);
	for (auto it : shapes)
	{
		builder.Add(compound, it);
	}
	// Create VTK actors for each processed shape
	vtkNew<IVtkTools_ShapeDataSource> occSource;
	occSource->SetShape(new IVtkOCC_Shape(compound));
	occSource->Update();
	vtkNew<vtkPolyDataMapper> mapper;
	mapper->SetInputConnection(occSource->GetOutputPort());
	vtkNew<vtkActor> actorSphere;
	actorSphere->SetMapper(mapper.Get());
	actorSphere->GetProperty()->SetColor(0.0, 0.0, 1.0);
        {
	//大括号限定锁的作用范围，确保锁的自动释放
	lock_guard<mutex> lock(mtx);
	render->AddActor(actorSphere.Get());
        }
}

```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by SaumyaSubham** [comment]
Response #1 I think to accelerate the rendering of VTK and the transformation of data from OCCT (Open CASCADE Technology) to VTK (Visualization Toolkit), several strategies can be employed. First, consider optimizing the data conversion process from OCCT to VTK. The current method involves creating a compound shape and processing it into a VTK data source in a single-threaded manner for each subset of shapes. Instead, parallelizing the conversion of each individual shape into VTK objects directly without combining them into a compound first could reduce the time complexity and improve performance. Additionally, minimizing the use of mutexes by reducing the critical section where rendering actors are added can also help, as excessive locking can cause significant delays in multi-threaded environments. Furthermore, employing VTK’s multi-threaded capabilities more effectively, such as using VTK's native support for parallel processing in the rendering pipeline (e.g., vtkMultiThreader or vtkSMPTools), can also speed up the rendering process. Another potential optimization is to leverage VTK's support for GPU-based rendering and data processing, which can significantly accelerate rendering performance, especially when dealing with a large number of objects like small spheres. Finally, reducing the geometric complexity of the spheres, if possible, or using lower-level representations until finer details are necessary, can also contribute to faster rendering times.

This might help you!

**Comment by newguyyy** [comment]
response #2
hello @SaumyaSubham Thank you for your help
I tried your suggestion. 1. For each Sphere TopoDS_Shape do not compose into TopoDS_Compound, directly convert into VTK data in each thread, add it to the renderer, so that the attempt will increase the total time, for the same number of Spheres, after compositing occSource->Update(); is less time-consuming.

2. Use VTK's built-in vtkMultiThreader for multi-threaded operation, but my VTK version is 6.1, and I encountered some errors when I tried, the data is empty after multi-threaded data transfer, which should be related to the multi-threaded function that VTK comes with
```
threader->SetNumberOfThreads(numTreads);

for (Standard_Integer i = 0; i < numTreads; i++)
{
    TopTools_SequenceOfShape subsets;
    ThreadData* data = new ThreadData();
    data->render = newRen.Get();
    for (Standard_Integer j = i * shapesPerthread; j < (i + 1) * shapesPerthread && j < SeqShape.Length(); j++)
    {
        //BRepMesh_IncrementalMesh meshshape(SeqShape(j + 1), deflection, relative);
        subsets.Append(SeqShape(j + 1));
    }
    data->SetShapes(subsets);
    cout << static_cast<void*>(data) << endl;
    threader->SetSingleMethod(ProcessShape, static_cast<void*>(data));
}
// 启动多线程执行
threader->SingleMethodExecute();
```
```
VTK_THREAD_RETURN_TYPE ProcessShape(void* arg)
{
	// 将 void* 参数转换为 ThreadData* 类型
	cout << "   " << arg << endl;
	ThreadData* data = static_cast<ThreadData*>(arg);

	//vector<vtkSmartPointer<vtkActor>> actors;
	TopoDS_Compound compound;
	BRep_Builder builder;
	builder.MakeCompound(compound);
	if (data->shapes.IsEmpty())
	{
		cout << "data is Empty" << endl;
	}
	cout << data->render << endl;

	for (auto it : data->shapes)
	{
		builder.Add(compound, it);
	}

	 //Create VTK actors for each processed shape
	IVtkTools_ShapeDataSource* occSource = IVtkTools_ShapeDataSource::New();
	occSource->SetShape(new IVtkOCC_Shape(compound));
	
	occSource->Update();

	vtkNew<vtkPolyDataMapper> mapper;
	mapper->SetInputConnection(occSource->GetOutputPort());
	vtkNew<vtkActor> actorSphere;
	actorSphere->SetMapper(mapper.Get());
	actorSphere->GetProperty()->SetColor(0.0, 0.0, 1.0);
	//cout << name << "  " << aftertime - beforetime << endl;

	// Collect actors to add them all at once later
	//actors.push_back(actorSphere.Get());
	
	//lock the mutex before adding the actor to the renderer
	{
		//大括号限定锁的作用范围，确保锁的自动释放
		lock_guard<mutex> lock(mtx);
		data->render->AddActor(actorSphere.Get());
		
	}
	// 清理内存
	delete data;
	return VTK_THREAD_RETURN_VALUE;
}
```
```class ThreadData {
public:
	
	//引用类型成员通常要求在对象创建时就进行初始化
	TopTools_SequenceOfShape shapes;
	vtkRenderer* render;
};
```
3. For mutex, I think I've minimized the use of mutex

Now I'm trying to read the updata function in the VTKalgorithm source code, figure out the specific principle of the updata function, and then optimize it

and you can read my repostory which contains my source code
link:[https://github.com/newguyyy/makeShperes](url)
